### PR TITLE
Icmp v4 and v6 - working echo request/reply and more to come

### PR DESCRIPTION
**Pull request by robs-zeynet**
This is still a work in progress and needs some cleanup (e.g., refactoring to reduce code duplication between v4 and v6), but some useful features are starting to work.  Apologies for the monster patches, but a lot of the code seems to be coupled (e.g., adjust an enum definition --> update all of the places it's referenced), so I'm open to opinions about how to structure it.

I went with this balance between 'fat struct' vs. 'nested enums' - see the discussion I tried to start with #18 for why this isn't an easy trade-off.

```
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum TransportHeader {
 +  Icmp4(icmp4::Icmp4Header),  // treat icmp4 as a whole different transport proto from icmp6
 +  Icmp6(icmp6::Icmp6Header),
     Udp(udp::UdpHeader),
    Tcp(tcp::TcpHeader)

}
```

Things that work:

1) Marshalling/unmarshalling basic IPv4 Echo Request packets (with tests!)
2) The scaffolding to add (all?) other Icmp4 and Icmp6 types (v6 Route Advertisement might be annoying)

```
let builder = PacketBuilder::
      ipv4(   [192,168,1,1],  //source ip
                [192,168,1,2], //destination ip
                20)            //time to life
      .icmp4_echo_request(1, 2);   // create an EchoRequest packet with id number = 1, and sequence number = 2
```

Some of the most basic tests are passing but I've also broken something in packet_compositions::test_compositions that needs to get fixed still.  In any case, this seems like there's enough here to let folks start looking at it.

Feedback welcome but I'll keep working on this and adding tests/functionality as I get time.

### CODE REVIEW

File: src/transport/icmp4.rs
Comment (JulianSchmid) [inline src/transport/icmp4.rs:228]: I would directly return the u8.

In the past I also always tried to setup some enum values, but it was ultimately more trouble then it was worth. Keeping an enum up-to-date is almost impossible and then there is the chance that a valid packet gets rejected because of an unknown enum value. It is less troublesome to allow all values for which we are not 100% sure if they are invalid.

Then there is no possible error case and the u8 can be returned without being wrapped in a Result (+ inline).

```suggestion
    #[inline]
    pub fn icmp_type(&self) -> u8 {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of IcmpV4Header::SERIALIZED_SIZE (8).
        unsafe {
            *self.slice.get_unchecked(0)
        }
    }
```

File: src/transport/icmp6.rs
Comment (JulianSchmid) [inline src/transport/icmp6.rs:146]: Flag as inline, correct comment & directly return value.

```suggestion
    pub fn icmp_chksum(&self) -> u16 {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of IcmpV6Header::SERIALIZED_SIZE (8).
        unsafe {
            get_unchecked_be_u16(self.slice.as_ptr().add(2))
        }
    }
```

File: src/transport/icmp4.rs
Comment (JulianSchmid) [inline src/transport/icmp4.rs:233]: Directly return the value & flag as inline.

```suggestion
    pub fn icmp_code(&self) -> Result<u8, ValueError> {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of IcmpV4Header::SERIALIZED_SIZE (8).
        unsafe {
            *self.slice.get_unchecked(1)
        }
    }
```

File: src/transport/icmp4.rs
Comment (JulianSchmid) [inline src/transport/icmp4.rs:242]: Directly return the value & flag as inline.

```suggestion
    pub fn icmp_chksum(&self) -> u16 {
        // SAFETY:
        // Safe as the contructor checks that the slice has
        // at least the length of UdpHeader::SERIALIZED_SIZE (8).
        unsafe {
            get_unchecked_be_u16(self.slice.as_ptr().add(2))
        }
    }
```

File: src/packet_slicing.rs
Comment (JulianSchmid) [inline src/packet_slicing.rs:16]: 👍

File: src/transport/icmp4.rs
Comment (JulianSchmid) [inline src/transport/icmp4.rs:27]: I am not sure if it is worth it to try to build an abstraction between IcmpV4 & IcmpV6.

I would leave the trait out in the first implementation and let the user abstract over the two protocols on their own if they need it.

File: src/packet_builder.rs
Comment (JulianSchmid) [inline src/packet_builder.rs:494]: I would remove the echo from the next step and instead add four methods for the echo case directly:

* `icmp4_echo_reply(seq, id)`
* `icmp6_echo_request(seq, id)`
* `icmp6_echo_reply(seq, id)`
* `icmp6_echo_reply(seq, id)`

that way it is no needed to specify the type & code as they are implicitly known from the type (also plays nicely with the `content` comment I wrote for the IcmpV4Header & IcmpV6Header). You also don't need a sanity check.

But I would keep the `icmp4` & `icmp6` methods so it is possible to write arbitrary icmp packets (if the user is so inclined).

Then creating an ICMP echo replay packet would then look like this:

```
        let builder = PacketBuilder::
             ipv4(   [192,168,1,1],  //source ip
                 [192,168,1,2], //desitionation ip
                 20)            //time to life
              .icmp4_echo_request(1, 2)
```

**Review by JulianSchmid — CHANGES_REQUESTED** [review]
First: Thanks again for the PR.

I added a review to your PR and as I said in the issue ICMP is a beast. I am also not 100% sure what the best way is. Especially for the slices (maybe we should make enums out of those too?).

**Comment by robs-zeynet** [comment]
Thanks @JulianSchmid  for the detailed feedback!  I like the idea of merging the type and codes into a single enum, moving the type-specific data into the enum (I'll probably keep a helper function in the base, similar to UdpHeader) and your proposed changes to the packet builder.  There's also a long tail of other comments that seem to all make sense to me.  Let me try to do a pass on your comments and I'll update this PR and we can iterate.  

One question I did have - there's a fair bit of boiler plate code that we could reduce by adding a few dependencies.

**Comment by robs-zeynet** [comment]
@JulianSchmid - pushed a new commit with what I believe are all(?) of your suggested changes.  A few notes:

1) I only updated ipv4 - will update v6 if you agree with the direction
2) I left the previous git commits in place; we can squash them once we're on the same page for the direction (or just before merge), per your preference.
3) Let me know what other feedback you have but IMHO this is looks good directionally.

Thank you!

**Comment by JulianSchmid** [comment]
> @JulianSchmid - pushed a new commit with what I believe are all(?) of your suggested changes. A few notes:
> 
> 1. I only updated ipv4 - will update v6 if you agree with the direction

Looks good to me 👍 

> 2. I left the previous git commits in place; we can squash them once we're on the same page for the direction (or just before merge), per your preference.

That is also ok with me.

> 3. Let me know what other feedback you have but IMHO this is looks good directionally.
> 
> Thank you!

**Comment by robs-zeynet** [comment]
FYI: added ICMP6 support for echo request/reply and fixed 'cargo test' so it passes now.  Starting to make some progress.  @JulianSchmid  - let me know where you think the bar is for starting to merge what I have (just echo request/reply support) and I can start to work towards that while I add support for Dst unreachable.

**Comment by JulianSchmid** [comment]
Hi @robs-zeynet ,

Looks good to me. I will add a few commits with a few changes on top after the merge that I already have locally.

If you are ok with a merge please remove the "work in progress; do not merge" from the title of this PR.

Greets
Julian

**Comment by robs-zeynet** [comment]
@JulianSchmid I discovered a bug in the new checksum routines that I need to fix but I haven't root-caused it yet.  I'll fixup the title once that's resolved.  Any opinions about the commit structure?  It's a bit of a mess now, so the options are:

1) Leave it that way
2) Squash it all
3) Go back and rebase to make it look pretty

I'm leaning towards 1 or 2 - but let me know.  Also, FYI - I know this is close but I don't know how much time this week I'll have to work on it.

**Comment by robs-zeynet** [comment]
Found the checksum bugs (plural!) and fixed with tests.  Good to merge from my side except for any commit restructuring which in theory you could hand if you wanted.  Thank you @JulianSchmid !

**Comment by JulianSchmid** [comment]
Hi @robs-zeynet , no need to re-write the history. I will just work on top of it and merge it as soon as I am done. + Thanks for the bugfix.

File: README.md
Comment (robs-zeynet) [inline README.md:196]: Maybe also add rfc4443 (ICMP6)?

**Comment by robs-zeynet** [comment]
Let me know when you're done with changes- I've got a stack of diffs with some more tests and fleshing out the rest of destination unreachable and ttl exceeded.

**Comment by JulianSchmid** [comment]
@robs-zeynet I will not work on it for at least 10 hours. If you have something ready just push it on top.

**Comment by JulianSchmid** [comment]
Hi @robs-zeynet ,

I would give it another go other the weekend to merge your PR (meaning adding some stuff on top). Is there something you want to push on top before I do another round?

I was thinking on maybe making the header length (in bytes) variable based on the ICMP packet. That way we could include some more of the static fields in other kinds of ICMP messages (e.g. the timestamps in ICMPv4). 

Greets
Julian

**Comment by robs-zeynet** [comment]
I don't have anything more to merge at this time - what's here unblocks my use-case.  Thanks.

As for variable length messages, I agree and that's why I added the header_len() function  (see link below) as a place holder for this.

https://github.com/JulianSchmid/etherparse/pull/22/files#diff-4dba2fcaec519af2ea77293bb0afa4b504773dcc9a594fff896e067275ad23e3R836

FWIW, I don't have a lot of spare time, but if you let me know what changes you want to add, I can try to help.  

Thanks again.

**Comment by JulianSchmid** [comment]
Good point. I think I will rename it again to byte5to8 . Naming is  truly one of the hardest things in programming 😅

**Review by JulianSchmid — APPROVED** [review]

**Comment by robs-zeynet** [comment]
Yay!

**Pull request merged**
